django_emarsys/api.py

```diff
@@ -28,3 +35,109 @@ def trigger_event(event_id, email, context):
+            for field in response}
+
+
+def _transform_contact_data(contact):
+    return {FIELD_IDS_BY_NAME[k]: v for k, v in contact.items()}
+
+
+def _create_contacts(contacts):
+    """
+    Returns
+    (number_of_successful_creates,
+     {'name@domain.org': {'error_code': 'Error message'}})
+    """
+    assert len(contacts) <= BATCH_SIZE
+    result = client.call('/api/v2/contact', 'PUT', {'contacts': contacts})
+    return len(result['ids']), result['errors']
+
+
+def _update_contacts(contacts):
+    """
+    Returns
+    (number_of_successful_updates,
+     {'name@domain.org': {'error_code': 'Error message'}})
+    """
+    assert len(contacts) <= BATCH_SIZE
+    result = client.call('/api/v2/contact', 'PUT', {'contacts': contacts})
+    return len(result['ids']), result['errors']
+
+
+def get_contact_data(email):
+    return client.call('/api/v2/contact/getdata', 'POST',
+                       {'keyId': '3', 'keyValues': [email]})
+
+
+def sync_contacts(contacts, create_missing=True, create_only_fields=None):
+    """
+    contacts is a list of dictionaries like this:
+        [{
+            u'E-Mail': u'total-berlin-admin@total.de',
+            u'Gender': 2,
+            u'First Name': u'Admin',
+            u'Last Name': u'von Total Berlin',
+            ...
+        }, ...]
+
+    The dictionary keys are mapped to emarsys field ids using
+    settings.EMARSYS_FIELDS, which can be generated with `get_fields()`.
+    """
+
+    def chunked(it, n):
+        """
+        From http://stackoverflow.com/a/8991553
+        """
+        while True:
+            chunk = tuple(slice(it, n))
+            if not chunk:
+                return
+            yield chunk
+
+    if not create_only_fields:
+        create_only_fields = []
+
+    total_updated = 0
+    total_created = 0
+    missing_contacts = []
+    failed_contacts = []
+
+    contacts = map(_transform_contact_data, contacts)
+
+    if settings.EMARSYS_RECIPIENT_WHITELIST is not None:
+        contacts = filter(lambda contact: contact['3']  # 3=email
+                          in settings.EMARSYS_RECIPIENT_WHITELIST)
+
+    for chunk_of_contacts in chunked(contacts, BATCH_SIZE):
+        num_successful, errors = _update_contacts(chunk_of_contacts)
+        total_updated += num_successful
+        missing_contacts.extend(email
+                                for email, error_dict in errors
+                                if '2008' in error_dict)
+        failed_contacts.extend((email, error_dict)
+                               for email, error_dict in errors
+                               if '2008' not in error_dict)
+
+    if create_missing:
+        create_only_field_ids = [FIELD_IDS_BY_NAME[field_name] for field_name
+                                 in create_only_fields]
+        missing_contacts = [{k: v for k, v in contact
+                             if k not in create_only_field_ids}
+                            for contact in missing_contacts]
```

Comment:
contacts should be filtered with create_only_fields when updating, not when creating.